app/src/main/java/eu/kanade/tachiyomi/source/LocalSource.kt

```diff
@@ -154,9 +160,115 @@ class LocalSource(
                 }
             }
 
+        // get Metadata from ComicInfo.xml Files
+        val metadataRegex = "(no)?xml".toRegex()
+        val metadataFile = getMangaDirsFiles(manga.url, baseDirsFile)
+            .firstOrNull { it.extension.matches(metadataRegex) }
+
+        try {
+            val encoding = "UTF-8"
+            if (metadataFile?.extension == "xml") {
+                val comicInfo = xml.decodeFromReader<ComicInfo>(AndroidXmlReader(metadataFile.inputStream(), encoding))
+                if (comicInfo.title != null || comicInfo.writer != null || comicInfo.summary != null ||
+                    comicInfo.genre != null || comicInfo.penciller != null
+                ) setMetadataXml(comicInfo, manga)
+            } else if (metadataFile == null) {
+                var createNoXmlFile = true
+                val mangaDir = getMangaDir(manga.url, baseDirsFile)
+                val folderPath = mangaDir?.absolutePath
+
+                val chapterArchives = getMangaDirsFiles(manga.url, baseDirsFile)
+                    .filter { isSupportedFile(it.extension) && !it.extension.equals("epub", true) }
+                    .toList()
+
+                for (chapter in chapterArchives) {
+                    var breaker = false
+                    if (chapter.extension.equals("zip", true) || chapter.extension.equals("cbz", true)) {
+                        withIOContext {
+                            ZipFile(chapter).use { zip: ZipFile ->
+                                val zipEntry = zip.getEntry("ComicInfo.xml")
+
+                                if (zipEntry != null) {
+                                    zip.getInputStream(zipEntry).use { archiveInputStream ->
+                                        val comicInfo = xml.decodeFromReader<ComicInfo>(AndroidXmlReader(archiveInputStream, encoding))
+                                        val archiveInputStreamOut = zip.getInputStream(zipEntry)
+                                        setAndCopyMetadata(comicInfo, archiveInputStreamOut, manga, folderPath)
+                                    }
+                                    createNoXmlFile = false
+                                    breaker = true
+                                }
+                            }
+                        }
+                        if (breaker) break
+                    } else if (chapter.extension.equals("rar", true) || chapter.extension.equals("cbr", true)) {
+                        withIOContext {
+                            Archive(chapter).use { archive: Archive ->
+                                var comicInfoFileHeader: FileHeader? = null
+                                archive.fileHeaders.map {
+                                    if (it.fileName == "ComicInfo.xml") comicInfoFileHeader = it
+                                }
```

Comment:
Isn't this trying to find the first file called `ComicInfo.xml`

```suggestion
                                var comicInfoFileHeader = archive.fileHeaders.firstOrNull { it.fileName == "ComicInfo.xml" }
```